### PR TITLE
refactor: redesign `StringMatchFiniteAutomata`

### PR DESCRIPTION
**Pull request by samuelfac**
Related to #5164

- [X] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Java/blob/master/CONTRIBUTING.md).
- [X] This pull request is all my own work -- I have not plagiarized it.
- [X] All filenames are in PascalCase.
- [X] All functions and variable names follow Java naming conventions.
- [X] All new algorithms have a URL in their comments that points to Wikipedia or other similar explanations.
- [X] All new code is formatted with `clang-format -i --style=file path/to/your/file.java`

### CODE REVIEW

**Comment by vil02** [comment]
Let's focus on #5221 first.

File: src/main/java/com/thealgorithms/others/StringMatchFiniteAutomata.java [hunk 1]
```diff
@@ -1,80 +1,118 @@
...
 
-        searchPat(text, pat);
+            System.out.println("Enter pattern:");
+            String pattern = scanner.nextLine();
 
-        scanner.close();
+            Set<Integer> indexFound = searchPattern(text, pattern);
+            indexFound.forEach(System.out::println);
+        }
     }
```
Comment (vil02) [inline src/main/java/com/thealgorithms/others/StringMatchFiniteAutomata.java:37]: The `main` method is redundant.

File: src/main/java/com/thealgorithms/others/StringMatchFiniteAutomata.java [hunk 2]
```diff
@@ -1,80 +1,118 @@
...
+            state = finiteAutomata[state][text.charAt(i)];
 
-            if (state == m) {
-                System.out.println("Pattern found at index " + (i - m + 1));
+            if (state == patternLength) {
+                indexFound.add(i - patternLength + 1);
             }
         }
+        return indexFound;
     }
```
Comment (vil02) [inline src/main/java/com/thealgorithms/others/StringMatchFiniteAutomata.java:67]: I think that this method should look somehow like that:
```java
public static Set<Integer> searchPattern(final String text, final String pattern) {
    final var stateTransitionTable = computeStateTransitionTable(pattern);
    FiniteAutomata finiteAutomata = new FiniteAutomata(stateTransitionTable);

    Set<Integer> indexFound = new TreeSet<>();
    for (int i = 0; i < text.length(); i++) {
        finiteAutomata.consume(text.charAt(i));

        if (finiteAutomata.getState() == pattern.length()) {
            indexFound.add(i - pattern.length() + 1);
        }
    }
    return indexFound;
}
```
The `FiniteAutomata` is a class with two `private` fields: `state` and `stateTransitionTable`. The method
`public consume(final char input)` updates the `state` according to `stateTransitionTable`.

**Review by vil02 — CHANGES_REQUESTED** [review]

File: src/main/java/com/thealgorithms/others/StringMatchFiniteAutomata.java [hunk 3]
```diff
@@ -1,80 +1,126 @@
...
-    public static void main(String[] args) {
-        scanner = new Scanner(System.in);
-        System.out.println("Enter String");
-        String text = scanner.nextLine();
-        System.out.println("Enter pattern");
-        String pat = scanner.nextLine();
 
-        searchPat(text, pat);
+    // Constants
+    private static final int CHARS = 256; // Total number of characters in the input alphabet
```
Comment (vil02) [inline src/main/java/com/thealgorithms/others/StringMatchFiniteAutomata.java:14]:
```suggestion
    private static final int CHARS = Character.MAX_VALUE + 1; // Total number of characters in the input alphabet
```
`char` has 16 bits in java. Therefore maybe one could consider using a `Map`-like structure to represent the state transition table.

**Review by vil02 — CHANGES_REQUESTED** [review]

File: src/test/java/com/thealgorithms/others/StringMatchFiniteAutomataTest.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+    @ParameterizedTest
+    @MethodSource("provideTestCases")
+    void searchPattern(String text, String pattern, Set<Integer> expectedOutput) {
+        assertEquals(expectedOutput, StringMatchFiniteAutomata.searchPattern(text, pattern));
+    }
+
+    private static Stream<Arguments> provideTestCases() {
+        return Stream.of(Arguments.of("abcbcabc", "abc", Set.of(0, 5)), //
+            Arguments.of("", "abc", Set.of()), //
```
[the remark this answers is not in the document]
Comment (vil02) [reply src/test/java/com/thealgorithms/others/StringMatchFiniteAutomataTest.java:21]: Well... I would leave formatting to clang-format and its setup [`.clang-format`](https://github.com/TheAlgorithms/Java/blob/cdb3affdd9ac4b8c7ccf4c55b05f6914bb9e73d6/.clang-format). I fully agree that having each test case in a separate line is more readable. But I would not _fight_ against clang-format by adding `//`. I am sure that it is possible to change the config of it in such a way, that these test cases will appear in different lines.

[the remark this answers is not in the document]
Comment (vil02) [reply src/test/java/com/thealgorithms/others/StringMatchFiniteAutomataTest.java]: Please add the suggested test cases. I think that they are quite useful edge cases.

File: src/test/java/com/thealgorithms/others/StringMatchFiniteAutomataTest.java [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+    @ParameterizedTest
+    @MethodSource("provideTestCases")
+    void searchPattern(String text, String pattern, Set<Integer> expectedOutput) {
+        assertEquals(expectedOutput, StringMatchFiniteAutomata.searchPattern(text, pattern));
+    }
+
+    private static Stream<Arguments> provideTestCases() {
+        return Stream.of(Arguments.of("abcbcabc", "abc", Set.of(0, 5)), Arguments.of("", "abc", Set.of()), Arguments.of("abcdefg", "xyz", Set.of()), Arguments.of("abcde", "", Set.of(1, 2, 3, 4, 5)), Arguments.of("abcabcabc", "abc", Set.of(0, 3, 6)), Arguments.of("abcabcabc", "abcabcabc", Set.of(0)),
+            Arguments.of("aaabbbaaa", "aaa", Set.of(0, 6)), Arguments.of("abcdefg", "efg", Set.of(4)));
```
Comment (vil02) [inline src/test/java/com/thealgorithms/others/StringMatchFiniteAutomataTest.java:21]:
```suggestion
                return Stream.of(Arguments.of("abcbcabc", "abc", Set.of(0, 5)), Arguments.of("", "abc", Set.of()), Arguments.of("", "", Set.of()), Arguments.of("a", "b", Set.of()), Arguments.of("a", "a", Set.of(0)), Arguments.of("abcdabcabcabcd", "abcd", Set.of(0, 10)), Arguments.of("abc", "bcd", Set.of()),
            Arguments.of("abcdefg", "xyz", Set.of()), Arguments.of("abcde", "", Set.of(1, 2, 3, 4, 5)), Arguments.of("abcabcabc", "abc", Set.of(0, 3, 6)), Arguments.of("abcabcabc", "abcabcabc", Set.of(0)), Arguments.of("aaabbbaaa", "aaa", Set.of(0, 6)), Arguments.of("abcdefg", "efg", Set.of(4)));
```

**Review by vil02 — APPROVED** [review]
